src/util/address.rs

```diff
@@ -523,23 +531,23 @@ impl Address {
 
     /// Get the address type of the address.
     /// None if unknown, non-standard or related to the future witness version.
-    pub fn address_type(&self) -> Option<AddressType> {
+    pub fn address_type(&self) -> Result<AddressType, Error> {
         match self.payload {
-            Payload::PubkeyHash(_) => Some(AddressType::P2pkh),
-            Payload::ScriptHash(_) => Some(AddressType::P2sh),
+            Payload::PubkeyHash(_) => Ok(AddressType::P2pkh),
+            Payload::ScriptHash(_) => Ok(AddressType::P2sh),
             Payload::WitnessProgram {
                 version,
                 program: ref prog,
             } => {
                 // BIP-141 p2wpkh or p2wsh addresses.
                 match version {
                     WitnessVersion::V0 => match prog.len() {
-                        20 => Some(AddressType::P2wpkh),
-                        32 => Some(AddressType::P2wsh),
-                        _ => None,
+                        20 => Ok(AddressType::P2wpkh),
+                        32 => Ok(AddressType::P2wsh),
+                        len => Err(Error::InvalidWitnessProgramLength(len)),
                     },
-                    WitnessVersion::V1 if prog.len() == 32 => Some(AddressType::P2tr),
-                    _ => None,
+                    WitnessVersion::V1 if prog.len() == 32 => Ok(AddressType::P2tr),
+                    version => Err(Error::UnsupportedWitnessVersion(version)),
                 }
```

Comment:
`match (version, prog.len())` would look a bit nicer but not a big deal.

But thinking about it isn't `Address` supposed to guarantee that lengths are always valid? This is checked at parsing and I don't seen an alternative unchecked constructor. If I'm right then these checks should be unnecessary. If not, then maybe it'd be better to have (private) `fn check_program_len(self, len: usize) -> Result<(), Error>` method on `WitnessVersion` and use it in both places to DRY.

Reply:
I was thinking about this myself. However, one day somebody may introduce new methods to create an address without checks and without this error checking we will get problems. So I think it is better to have an unreachable error checking code rather than assume everything should be fine due to the code in other external methods and unwrap/expect.

Reply:
Right, maybe a more relevant question is if we should commit into address having this invariant.

Reply:
I think this is outside of the scope of this PR and may be controversial on how to achieve this